### PR TITLE
When using the `task` and `workflow` decorator, correctly wrap the fu…

### PR DESCRIPTION
**Pull request by bstadlbauer**
# TL;DR
This enables tooling such as docstring search tools to unwrap the object and show the correct docstring. The method sets the `__wrapped__` attribute of the created instance.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
We're using [mkdocs](https://www.mkdocs.org/) together with [mkdocstrings](https://mkdocstrings.github.io/usage/) to render our Python documentation. We noticed, that documentation for Flyte tasks didn't show up, and upon further investigation found that `mkdocstrings` unwraps all decorated elements to show the correct documentation. See also: https://mkdocstrings.github.io/troubleshooting/#my-wrapped-function-shows-documentationcode-for-its-wrapper-instead-of-its-own

For this unwrapping to work, wrapper elements need to have `__wrapped__` set to a reference of the wrapped (decorated) function. With functions one could use [`functools.wraps()`](https://docs.python.org/3/library/functools.html#functools.wraps), but as Flyte is returning instances of Flyte objects, [`functools.update_wrapper()`](https://docs.python.org/3/library/functools.html#functools.update_wrapper) is what works.

### CODE REVIEW

**Comment by eapolinario** [comment]
@bstadlbauer , can you investigate the unit test failure? This [test](https://github.com/flyteorg/flytekit/blob/f500af0dd7d8a1310f8819ebc2152e21fb458f4f/plugins/flytekit-greatexpectations/tests/test_task.py#L133) is failing and it seems that it might be a weird interaction with the wrapped function and the pytest.mark.xfail. We should be able replace that use with the more idiomatic `with pytest.raises` (e.g. https://github.com/flyteorg/flytekit/blob/8528268a29a07fe7e9ce9f7f08fea68c41b6a60b/tests/flytekit/unit/common_tests/exceptions/test_scopes.py#L18-L19).

**Comment by bstadlbauer** [comment]
Hi @eapolinario! 

Thank you for the fast reply. I've checked, and you're right. Seems like what actually should happen is a `with pytest.raises()` instead of the `pytest.mark.xfail`. I've updated the test accordingly

File: tests/flytekit/unit/core/test_wrapping.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from flytekit import task, workflow
+
+
+def test_task_correctly_wrapped():
```
Comment (eapolinario) [inline tests/flytekit/unit/core/test_wrapping.py:4]: @bstadlbauer , we allow for stacking of @task decorators (e.g. https://github.com/flyteorg/flytekit/blob/master/tests/flytekit/unit/core/functools/stacked_decorators.py). Can you add a unit test to confirm the behavior too?

**Review by eapolinario — COMMENTED** [review]
@bstadlbauer , thanks for making this change, it looks good and should be ok to merge after we have that extra unit test I mentioned in the comment.

File: tests/flytekit/unit/core/test_wrapping.py (hunk 1 above)
Comment (bstadlbauer) [reply tests/flytekit/unit/core/test_wrapping.py:4]: @eapolinario I've added the corresponding test, seems to work as expected. 

Also, I wasn't sure where to place the tests, so if you feel like there would be a better spot, just let me know.

Comment (eapolinario) [reply tests/flytekit/unit/core/test_wrapping.py:6]: leaving them here is fine.

**Review by eapolinario — APPROVED** [review]
